python/lsst/ts/dream/csc/dream_csc.py

```diff
@@ -479,6 +530,116 @@ async def send_events(self, status_data: dict[str, Any]) -> None:
+        other = (
+            f"{data_product.start.tai.isot}_{data_product.server}_"
+            f"{data_product.kind}{product_type}_"
+            f"{data_product.seq[0]:06d}_{data_product.seq[-1]:06d}"
+        )
+        key = self.s3bucket.make_key(
+            salname=self.salinfo.name,
+            salindexname=None,
+            generator="dream",
+            date=data_product.start,
+            other=other,
+            suffix=".fits",
+        )
+
+        # Download the object with HTTP
+        dream_url = f"{self.config.url_root}/{data_product.filename}"
+        async with httpx.AsyncClient() as client:
+            async with client.stream("GET", dream_url) as response:
+                response.raise_for_status()
+
+                try:
+                    # First attempt: Save to S3
+                    await self.save_to_s3(response, key)
+                    return  # Success!
+                except Exception:
+                    self.log.exception(
+                        f"Could not upload {key} to S3; trying to save to local disk."
+                    )
+                    await self.save_to_local_disk(response, key)
+
+    async def save_to_s3(self, response: httpx.Response, key: str) -> None:
+        """Upload file to S3 from an HTTP stream.
+
+        Parameters
+        ----------
+        response : `httpx.Response`
+            HTTP response object for the file to save.
+
+        key : `str`
+            S3 style filename key to save to.
+        """
+        if not self.s3bucket:
+            raise RuntimeError("S3 bucket not configured")
+
+        with io.BytesIO(await response.aread()) as buffer:
+            await self.s3bucket.upload(fileobj=buffer, key=key)
+        url = (
+            f"{self.s3bucket.service_resource.meta.client.meta.endpoint_url}/"
+            f"{self.s3bucket.name}/{key}"
+        )
+        await self.evt_largeFileObjectAvailable.set_write(
+            url=url,
+            generator="dream",
+        )
+        self.log.info(f"Successfully uploaded {key} to S3.")
+
+    async def save_to_local_disk(self, response: httpx.Response, key: str) -> None:
+        """Save the file from an HTTP stream to local disk.
+
+        Parameters
+        ----------
+        response : `httpx.Response`
+            HTTP response object for the file to save.
+
+        key: `str`
+            S3 style filename key to save to.
+        """
+        if not self.s3bucket:
+            raise RuntimeError("S3 bucket not configured")
+
+        filepath = pathlib.Path("/tmp") / self.s3bucket.name / key
+        dirpath = filepath.parent
+        if not dirpath.exists():
+            self.log.info(f"Creating directory {str(dirpath)}")
+            dirpath.mkdir(parents=True, exist_ok=True)
+
+        with open(filepath, "wb") as tmp_file:
+            async for chunk in response.aiter_bytes():
+                tmp_file.write(chunk)
+
+        await self.evt_largeFileObjectAvailable.set_write(
```

Comment:
So, this event is reserved from when we write things to the S3 bucket so, in this case, don't publish it.